homeassistant/components/zha/core/gateway.py

```diff
@@ -166,6 +177,25 @@ def __init__(self, hass, config, config_entry):
         )
         async_dispatcher_send(self._hass, SIGNAL_ADD_ENTITIES)
 
+    @staticmethod
+    async def async_options_updated(hass, config_entry):
+        """Handle the update signal from updating the config entry."""
+        gateway = hass.data[DATA_ZHA][DATA_ZHA_GATEWAY]
+        await gateway.async_update_options(config_entry)
+
+    async def async_update_options(self, config_entry):
+        """Triggered by config entry options updates."""
+        _LOGGER.debug(
+            "ZHA config entry options updated from: %s to: %s",
+            self._config_entry.options,
+            config_entry.options,
+        )
+        self._config_entry = config_entry
+        self.application_controller.use_source_routing = self._config_entry.options.get(
```

Comment:
Hrm, currently the radio libs do not support live changes to source routing. Lets keep this static for now and just pass `config={"source_routing": true}` dict into https://github.com/home-assistant/home-assistant/blob/6852ccd8de66d7cf647f609b437aa87c9a9545c8/homeassistant/components/zha/core/gateway.py#L132 prior the zigpy is started.

Reply:
It works fine for me... I’ll attach logs that show it.

Reply:
on a side note, if we keep interface between `homeassistant-core` and zigpy radios down to `config` dict, then we can get rid off `await radio.connect()` in ZHA core and just pass the serial path in the same config. This would allow us to limit the initialization interface down to `startup()` and `shutdown()` methods.

Reply:
```
2020-02-28 07:39:44 DEBUG (MainThread) [bellows.ezsp] Send command nop: ()
2020-02-28 07:39:44 DEBUG (bellows.thread_0) [bellows.uart] Sending: b'16cb21ad3f687e'
2020-02-28 07:39:44 DEBUG (bellows.thread_0) [bellows.uart] Data frame: b'62cba1adac747e'
2020-02-28 07:39:44 DEBUG (bellows.thread_0) [bellows.uart] Sending: b'87009f7e'
2020-02-28 07:39:44 DEBUG (MainThread) [bellows.ezsp] Application frame 5 (nop) received: b''
2020-02-28 07:39:53 DEBUG (MainThread) [homeassistant.components.zha.core.gateway] ZHA config entry options updated from: {'enable_source_routing': False, 'radio_type': 'ezsp', 'usb_path': '/dev/tty.GoControl_zigbee'} to: {'enable_source_routing': False, 'radio_type': 'ezsp', 'usb_path': '/dev/tty.GoControl_zigbee'}
2020-02-28 07:39:53 DEBUG (MainThread) [bellows.ezsp] Send command setConcentrator: (False, <EmberConcentratorType.HIGH_RAM_CONCENTRATOR: 65529>, 600, 1800, 2, 5, 0)
2020-02-28 07:39:53 DEBUG (bellows.thread_0) [bellows.uart] Sending: b'27c821b854d3eaea5b9c4d27af558b747e'
2020-02-28 07:39:53 DEBUG (bellows.thread_0) [bellows.uart] Data frame: b'73c8a1b854d3467e'
2020-02-28 07:39:53 DEBUG (bellows.thread_0) [bellows.uart] Sending: b'8070787e'
2020-02-28 07:39:53 DEBUG (MainThread) [bellows.ezsp] Application frame 16 (setConcentrator) received: b'00'
2020-02-28 07:39:53 DEBUG (MainThread) [bellows.zigbee.application] Set concentrator type: [<EmberStatus.SUCCESS: 0>]
2020-02-28 07:39:54 DEBUG (MainThread) [bellows.ezsp] Send command nop: ()
2020-02-28 07:39:54 DEBUG (bellows.thread_0) [bellows.uart] Sending: b'30c921ad41df7e'
2020-02-28 07:39:54 DEBUG (bellows.thread_0) [bellows.uart] Data frame: b'04c9a1adbc5f7e'
2020-02-28 07:39:54 DEBUG (bellows.thread_0) [bellows.uart] Sending: b'8160597e'
2020-02-28 07:39:54 DEBUG (MainThread) [bellows.ezsp] Application frame 5 (nop) received: b''
2020-02-28 07:39:55 DEBUG (MainThread) [homeassistant.components.zha.core.channels.base] [0xe1b9:1:0x0b04]: async_update
2020-02-28 07:39:55 DEBUG (MainThread) [homeassistant.components.zha.core.channels.base] [0x710d:1:0x0b04]: async_update
2020-02-28 07:39:55 DEBUG (MainThread) [homeassistant.components.zha.core.channels.base] [0x7359:1:0x0b04]: async_update
2020-02-28 07:40:04 DEBUG (MainThread) [bellows.ezsp] Send command nop: ()
2020-02-28 07:40:04 DEBUG (bellows.thread_0) [bellows.uart] Sending: b'41ce21adf08e7e'
2020-02-28 07:40:04 DEBUG (bellows.thread_0) [bellows.uart] Data frame: b'15cea1ad54dc7e'
2020-02-28 07:40:04 DEBUG (bellows.thread_0) [bellows.uart] Sending: b'82503a7e'
2020-02-28 07:40:04 DEBUG (MainThread) [bellows.ezsp] Application frame 5 (nop) received: b''
2020-02-28 07:40:14 DEBUG (MainThread) [bellows.ezsp] Send command nop: ()
2020-02-28 07:40:14 DEBUG (bellows.thread_0) [bellows.uart] Sending: b'52cf21ad47c57e'
2020-02-28 07:40:14 DEBUG (bellows.thread_0) [bellows.uart] Data frame: b'26cfa1add4d97e'
2020-02-28 07:40:14 DEBUG (bellows.thread_0) [bellows.uart] Sending: b'83401b7e'
2020-02-28 07:40:14 DEBUG (MainThread) [bellows.ezsp] Application frame 5 (nop) received: b''
2020-02-28 07:40:18 DEBUG (MainThread) [homeassistant.components.zha.core.gateway] ZHA config entry options updated from: {'enable_source_routing': True, 'radio_type': 'ezsp', 'usb_path': '/dev/tty.GoControl_zigbee'} to: {'enable_source_routing': True, 'radio_type': 'ezsp', 'usb_path': '/dev/tty.GoControl_zigbee'}
2020-02-28 07:40:18 DEBUG (MainThread) [bellows.ezsp] Send command setConcentrator: (True, <EmberConcentratorType.HIGH_RAM_CONCENTRATOR: 65529>, 600, 1800, 2, 5, 0)
2020-02-28 07:40:18 DEBUG (bellows.thread_0) [bellows.uart] Sending: b'63cc21b855d3eaea5b9c4d27af5537867e'
2020-02-28 07:40:18 DEBUG (bellows.thread_0) [bellows.uart] Data frame: b'37cca1b85481d97e'
2020-02-28 07:40:18 DEBUG (bellows.thread_0) [bellows.uart] Sending: b'8430fc7e'
2020-02-28 07:40:18 DEBUG (MainThread) [bellows.ezsp] Application frame 16 (setConcentrator) received: b'00'
2020-02-28 07:40:18 DEBUG (MainThread) [bellows.zigbee.application] Set concentrator type: [<EmberStatus.SUCCESS: 0>]
```

Reply:
> It works fine for me

Yep, i took a second look and it should work, but it was unintentional :) My opinion we should  keep interface between ZHA and zigpy to a minimum and allow radio libs to pick keys from config dict accordingly to what library supports. should minimize changes in the future

Reply:
We should allow real-time changes. Forcing restarts is a terrible user experience.

Reply:
>so I'm not sure what you mean when you say live changes aren't supported... can you clarify?

I mean  i cannot guarantee it would work the same way for every radio. I don't have enough information atm, but some radios may need to have this setup prior the network is started. That's why i wanted to keep it static, i.e. config dict

Reply:
I don't disagree. But we're not there yet on zigpy side.
Let's do static config dict for now and when we have more information we can add something like `update_configuration()` or `restart()` to zigpy. 
and then we could also add "serial_port" and OTA configuration options to the same config dict and keep an unified interface.